### PR TITLE
Fix null-termination bug w/ `string`->`int` cast

### PR DESCRIPTION
**Pull request by jeremiah-corrado**
During AoC activities, we identified a bug involving a missing null-terminator in strings created by `IO.readLine`. The bug caused leftover bytes from previous `readLine` operations to show up in casts from `string` to `int`.

This PR modifies `IO.readStringBytesData` to properly terminate strings modified by `readLine(ref s: string)` and `readLine(ref b: bytes)`

It also ensures that string buffers are properly terminated before casts to integral types, by modifying `_cleanupForNumericCast`.

An additional test from AoC activities was added to confirm the validity of the fix.

- [x] passing paratest

### CODE REVIEW

**Comment by mppf** [comment]
> During AoC activities, we identified a bug involving a missing null-terminator in strings created by `IO.readLine`.

I was expecting that this PR would adjust IO.readLine to add a null terminator there somewhere but it does not. Why not?

**Comment by jeremiah-corrado** [comment]
Offline, @vasslitvinov suggested modifying `_cleanupForNumericCast` and it appears to have worked, so I decided to go with that.

I suppose adding a null terminator to readLine would be more correct tho? Maybe to protect against similar bugs other casts?

File: modules/internal/BytesStringCommon.chpl [hunk 1]
```diff
@@ -1426,6 +1426,9 @@ module BytesStringCommon {
         }
       }
     }
+
+    // ensure that there is a null byte at the end of the buffer
+    x.buff[x.buffLen] = 0;
```
Comment (mppf) [inline modules/internal/BytesStringCommon.chpl:1431]: Recommend `if x.buffLen > 0` here around this, unless you know that `x` cannot be the empty string at this point.

**Review by mppf — APPROVED** [review]

**Comment by bradcray** [comment]
Thanks so much for picking this up quickly @jeremiah-corrado!  Wish we'd discovered it prior to AoC 2022!

**Comment by jeremiah-corrado** [comment]
Sure thing, at least we got it early!

**Comment by bradcray** [comment]
True!  What if it hadn't been until day 10!  :O

**Pull request merged**